community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/pipes/CachingExpandInto.scala

```diff
@@ -84,26 +92,35 @@ trait CachingExpandInto {
   }
 
   private def relIterator(query: QueryContext, fromNode: Node,  toNode: Node, preserveDirection: Boolean,
-                          relTypes: Option[Seq[Int]], relCache: RelationshipsCache, dir: SemanticDirection) = {
+                          relTypes: Option[Seq[Int]], relCache: RelationshipsCache, dir: SemanticDirection): Iterator[Relationship] with AutoCloseable = {
     val (start, localDirection, end) = if(preserveDirection) (fromNode, dir, toNode) else (toNode, dir.reversed, fromNode)
     val relationships = query.getRelationshipsForIds(start, localDirection, relTypes)
-    new PrefetchingIterator[Relationship] {
-      //we do not expect two nodes to have many connecting relationships
-      val connectedRelationships = new ArrayBuffer[Relationship](2)
-
-      override def fetchNextOrNull(): Relationship = {
-        while (relationships.hasNext) {
-          val rel = relationships.next()
-          val other = rel.getOtherNode(start)
-          if (end == other) {
-            connectedRelationships.append(rel)
-            return rel
+    new Iterator[Relationship] with AutoCloseable
+    {
+      private val inner = new PrefetchingIterator[Relationship] {
```

Comment:
I know this is unrelated to your change, but should we depend on `PrefetchingIterator` in the cypher modules?